File: backend/routes/email.py
"""routes/email.py — Email Agent endpoints (read-only + approval-first send).

Every sensitive operation is gated by the existing permission system:
  - reading   → require_permission('email.read')    (mode: ask)
  - sending   → require_permission('email.send')    (mode: ask) + the message
                must exist as a server-side draft created via /draft
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from auth import require_boss
from ratelimit import is_rate_limited
from services import email_agent, permissions
from services.permissions import require_permission

router = APIRouter(prefix="/api/email", tags=["email"])


class DraftRequest(BaseModel):
    to: str
    subject: str = ""
    body: str = ""


class SendRequest(BaseModel):
    draft_id: str


def _handle_unavailable(exc: Exception):
    return HTTPException(status_code=503, detail=str(exc))


@router.get("/unread", dependencies=[Depends(require_boss), Depends(require_permission("email.read"))])
def email_unread_endpoint(limit: int = Query(15, ge=1, le=50)):
    """Recent unread emails (does NOT mark them read)."""
    try:
        return {"unread": email_agent.get_unread(limit=limit)}
    except email_agent.EmailUnavailableError as exc:
        raise _handle_unavailable(exc) from exc


@router.get("/summary", dependencies=[Depends(require_boss), Depends(require_permission("email.read"))])
def email_summary_endpoint():
    """Aggregated inbox summary (count, top senders, priority items)."""
    try:
        return {"summary": email_agent.summarize_inbox()}
    except email_agent.EmailUnavailableError as exc:
        raise _handle_unavailable(exc) from exc


@router.get("/search", dependencies=[Depends(require_boss), Depends(require_permission("email.read"))])
def email_search_endpoint(q: str = Query(..., min_length=1)):
    """Search inbox by subject/from."""
    try:
        return {"results": email_agent.search_emails(q, limit=10)}
    except email_agent.EmailUnavailableError as exc:
        raise _handle_unavailable(exc) from exc


@router.post("/draft", dependencies=[Depends(require_boss), Depends(require_permission("email.read"))])
def email_draft_endpoint(req: DraftRequest):
    """Create a server-side pending draft and return a preview for approval."""
    try:
        draft = email_agent.create_draft(req.to, req.subject, req.body)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except email_agent.EmailUnavailableError as exc:
        raise _handle_unavailable(exc) from exc
    return {
        "draft_id": draft["id"],
        "preview": {
            "to": draft["to"],
            "subject": draft["subject"],
            "body": draft["body"],
        },
        "expires_in_seconds": email_agent.DRAFT_TTL_SECONDS,
    }


@router.post("/send", dependencies=[Depends(require_boss), Depends(require_permission("email.send"))])
def email_send_endpoint(req: SendRequest):
    """Send a draft created via /draft (approval-first flow)."""
    try:
        result = email_agent.send_draft(req.draft_id)
    except email_agent.EmailUnavailableError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    permissions._audit("email.send", "allowed", f"sent to {result['to']}")
    return {"status": "ok", **result}


@router.post("/cancel", dependencies=[Depends(require_boss)])
def email_cancel_endpoint(req: SendRequest):
    """Discard a pending draft without sending."""
    ok = email_agent.cancel_draft(req.draft_id)
    return {"status": "ok" if ok else "not_found"}

File: friday-ui/src/api/email.js
import { API_BASE_URL } from './config.js';

const BASE = `${API_BASE_URL}/api/email`;

async function jsonOrThrow(res) {
  if (!res.ok) {
    let detail = `HTTP ${res.status}`;
    try {
      const body = await res.json();
      detail = typeof body.detail === 'string' ? body.detail
        : (body.detail?.message || JSON.stringify(body.detail));
    } catch (_) { /* keep default */ }
    throw new Error(detail);
  }
  return res.json();
}

/** Recent unread emails (server never marks them read). */
export async function fetchUnread(limit = 15) {
  const res = await fetch(`${BASE}/unread?limit=${limit}`);
  const data = await jsonOrThrow(res);
  return data.unread || [];
}

/** Aggregated inbox summary. */
export async function fetchEmailSummary() {
  const res = await fetch(`${BASE}/summary`);
  const data = await jsonOrThrow(res);
  return data.summary || null;
}

/** Search inbox by subject/from. */
export async function searchEmails(query) {
  const res = await fetch(`${BASE}/search?q=${encodeURIComponent(query)}`);
  const data = await jsonOrThrow(res);
  return data.results || [];
}

/** Create a server-side draft; returns { draft_id, preview, expires_in_seconds }. */
export async function createEmailDraft({ to, subject, body }) {
  const res = await fetch(`${BASE}/draft`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ to, subject, body }),
  });
  return jsonOrThrow(res);
}

/**
 * Approval-first send: grants a short-lived one-time approval for email.send
 * (the user just confirmed in the UI), then sends the server-side draft.
 */
export async function approveAndSendEmail(draftId) {
  const approveRes = await fetch(`${API_BASE_URL}/api/permissions/approve`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ capability: 'email.send', seconds: 180 }),
  });
  if (!approveRes.ok) {
    const body = await approveRes.json().catch(() => ({}));
    throw new Error(body.detail || 'Permission approval failed');
  }

  const res = await fetch(`${BASE}/send`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ draft_id: draftId }),
  });
  return jsonOrThrow(res);
}

/** Discard a pending draft. */
export async function cancelEmailDraft(draftId) {
  await fetch(`${BASE}/cancel`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ draft_id: draftId }),
  }).catch(() => {});
}
